commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/core/EuclideanNormAlgorithms.java

```diff
@@ -0,0 +1,751 @@
+                // Too small so scale up
+                scale = SCALE_UP;
+                rescale = SCALE_DOWN;
+            }
+
+            // Same as LinearCombination but with scaling.
+            // Splitting is safe due to scaling and only one term requires splitting.
+
+            // Implement dot2s (Algorithm 5.4) from Ogita et al (2005).
+            final int len = v.length;
+
+            // p is the standard scalar product sum.
+            // s is the sum of round-off parts.
+            double a = v[0] * scale;
+            double p = a * a;
+            double s = productLowUnscaled(a, p);
+
+            // Remaining split products added to the current sum and round-off sum.
+            for (int i = 1; i < len; i++) {
+                a = v[i] * scale;
+                final double h = a * a;
+                final double r = productLowUnscaled(a, h);
+
+                final double x = p + h;
+                // s_i = s_(i-1) + (q_i + r_i)
+                s += DoublePrecision.twoSumLow(p, h, x) + r;
+                p = x;
+            }
+            p += s;
+
+            return Math.sqrt(p) * rescale;
+        }
+
+        /**
+         * Compute the low part of the double length number {@code (z,zz)} for the exact
+         * square of {@code x} using Dekker's mult12 algorithm. The standard
+         * precision product {@code x*x} must be provided. The number {@code x}
+         * is split into high and low parts using Dekker's algorithm.
+         *
+         * <p>Warning: This method does not perform scaling in Dekker's split and large
+         * finite numbers can create NaN results.
+         *
+         * @param x The factor.
+         * @param xx Square of the factor (x * x).
+         * @return the low part of the product double length number
+         */
+        private static double productLowUnscaled(double x, double xx) {
+            // Split the numbers using Dekker's algorithm without scaling
+            final double hx = DoublePrecision.highPartUnscaled(x);
+            final double lx = x - hx;
+
+            return DoublePrecision.productLow(hx, lx, hx, lx, xx);
+        }
+    }
+
+    /** Modification of {@link ExtendedPrecisionLinearCombination} that only uses a single pass through
+     * the input array.
+     */
+    static final class ExtendedPrecisionLinearCombinationSinglePass implements ToDoubleFunction<double[]> {
+
+        /** Threshold for scaling small numbers. */
+        private static final double SMALL_THRESH = 0x1.0p-500;
+
+        /** Threshold for scaling large numbers. */
+        private static final double LARGE_THRESH = 0x1.0p+500;
+
+        /** Value used to scale down large numbers. */
+        private static final double SCALE_DOWN = 0x1.0p-600;
+
+        /** Value used to scale up small numbers. */
+        private static final double SCALE_UP = 0x1.0p+600;
+
+        /** {@inheritDoc} */
+        @Override
+        public double applyAsDouble(final double[] v) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 45 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 98% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 515-575 into a separate method.